### PR TITLE
Developer Experience Improvements

### PR DESCRIPTION
**Pull request by ConnorJC3**
**Is this a bug fix or adding new feature?**

Neither - it focuses on improvements to the developer experience.

**What is this PR about? / Why do we need it?**

This PR contains several improvements split up by commit:

*[Only collect metrics in Prow](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/commit/c01bda498f1310f17f71b9b48c32fa410e347578)*

Metrics collection runs locally by default, but this makes no sense because the metrics will fail to publish anyways. Check for the presence of `PROW_JOB_ID` and only run metrics in Prow jobs.

*[Simplify image building](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/commit/706174229b6e7d8b6b33375d657bb7c7e44170c1)*

Simplifies handling of image building in the `Makefile`. Removes some CI-only defaults from the `Makefile` that are confusing and difficult to use. Renames `hack/prow.sh` to `hack/cloudbuild.sh` to clearly indicate what it is used for.

[Split driver install/uninstall out to functions; Introduce make cluster/(un)install](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/commit/78c1cc47ce38d4898d1844d2d75225d7617ced61)

Introduces `make cluster/install` and `make cluster/uninstall` which can be used to (un)install the driver for local testing. In order to accomplish this, driver installation and uninstallation are split out into a separate function in `utils.sh`

[Cache GOCACHE and GOMODCACHE in Dockerfile](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/commit/9bc14dab49bff90a6e9aa8f063186274c51f8d8e)

Configures the `Dockerfile` to cache go modules and build output on supported platforms. In my local testing, this introduced an approximate 30x speedup (from 600 seconds to 20 seconds) when rebuilding the driver image with a small change (such as a dependency upgrade) to `go.mod`.

[Update Makefile documentation and CONTRIBUTING.md](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/commit/4d982c6e7d84526dcb74d638e819d273512858d0)

Makefile documentation was somewhat out of date, cleans it up. Also moves the quickstart guide to `CONTRIBUTING.md` and cleans up outdated information in `CONTRIBUTING.md`.

[Cleanup non-external e2e test handling](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/2079/commits/355f183848d19f13549d25c851be41a9fd91886d)

Cleans up the logic of non-external tests, fixing a bug that causes the tests to fail to run when run from inside a home directory that is a symlink.

**What testing is done?** 

CI/Manual

### CODE REVIEW

**Comment by AndrewSirenko** [comment]
Thanks for these **amazing** devloop + new contributor improvements. 

xref: #1929 (Because this PR resolves checklist items 1 & 3)

Working my way through testing the PR, eta end of week.

**Review by AndrewSirenko — COMMENTED** [review]
Thank you for this. It erases so many minutes from creating/reviewing PRs (except for the 10 minutes it took to download deps this PR because it's the first time a cache mount was  created 😅). 

Left mostly non-blocking comments except for adding a pre-requisite for having a private ECR repository named `aws-ebs-csi-driver` (or overriding that default) 

Besides ^^, I went through the new contributing guide steps and it all worked for me :)

**Review by torredil — APPROVED** [review]
/lgtm

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,46 +1,96 @@
...
+
+Export the `KUBECONFIG` environment variable to access the cluster for manual testing:
+```bash
+eval "$(make cluster/kubeconfig)"
+```
+
+To test a change, build the driver image and then install it:
+```bash
+make cluster/image
+make cluster/install
```
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply CONTRIBUTING.md:54]: Discussed offline and after some confusion we both realized this actually does work and no warning is needed

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,46 +1,96 @@
...
+```
+
+Export the `KUBECONFIG` environment variable to access the cluster for manual testing:
+```bash
+eval "$(make cluster/kubeconfig)"
+```
+
+To test a change, build the driver image and then install it:
+```bash
+make cluster/image
```
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply CONTRIBUTING.md:53]: `make cluster/image` will automatically create an ECR repo if `CREATE_MISSING_ECR_REPO` is `true` (which is its default value): https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/6015f5def653f1fd749bcd81fef4f61694cbad4f/hack/e2e/build-image.sh#L70-L75

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,46 +1,96 @@
...
-### Helm and manifests
-The Helm chart for this project is in the `charts/aws-ebs-csi-driver` directory.  The manifests for this project are in the `deploy/kubernetes` directory.  All of the manifests except kustomize patches are generated by running `helm template`.  This keeps the Helm chart and the manifests in sync.
+Run `make update` to automatically format go source files and re-generate automatically generated files. If `make update` produces any changes, commit them before submitting a PR.
 
-When updating the Helm chart:
-* Generate manifests: `make generate-kustomize`
-* There are values files in `deploy/kubernetes/values` used for generating some of the manifests
-* When adding a new resource template to the Helm chart please update the `generate-kustomize` make target, the `deploy/kubernetes/values` files, and the appropriate kustomization.yaml file(s).
\ No newline at end of file
+Run `make verify` to run linters and other similar code checking tools. Fix any issues `make verify` discovers before submitting a PR.
```
[the remark this answers is not in the document]
Comment (ConnorJC3) [reply CONTRIBUTING.md:96]: I think writing down exactly what to put in the PR is better suited for the PR template (which I have plans to revamp soon™), but I can add some generic language about testing helm changes if that's ok with you?

[the remark this answers is not in the document]
Comment (AndrewSirenko) [reply CONTRIBUTING.md:96]: Either waiting for next pr and/or adding generic language in this PR sounds great. Thank you.

File: CONTRIBUTING.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndrewSirenko) [reply CONTRIBUTING.md:53]: Today I learned. Thank you.

**Comment by AndrewSirenko** [comment]
/lgtm

Will approve once devloop with EKS confirmed works.

**Pull request merged**